File: examples/enterprise/python/ent_04_kafka_tumbling_window.py
"""Enterprise 04 · Kafka → 10-second tumbling window → console

Reads order events from Kafka, assigns them to 10-second tumbling windows
based on event timestamp (field "ts" as epoch ms), and emits per-window
aggregates: count, sum_amount. Late events (arriving > 2 s after the
watermark) are logged but still counted in their assigned window.

Prerequisites:
    make infra-up seed-kafka

Run:
    KAFKA_BOOTSTRAP=localhost:9092 python python/ent_04_kafka_tumbling_window.py
"""

import json
import os
import time
from collections import defaultdict

import pyarrow as pa
import pandas as pd
from confluent_kafka import Consumer, KafkaException

BOOTSTRAP = os.environ.get("KAFKA_BOOTSTRAP", "localhost:9092")
WINDOW_MS = 10_000
WATERMARK_LAG_MS = 2_000
TIMEOUT_S = 20

SYNTHETIC_EVENTS = [
    {"customer": "alice", "amount": 120.0,  "ts": 1_716_200_000_000},
    {"customer": "bob",   "amount": 45.0,   "ts": 1_716_200_001_000},
    {"customer": "carol", "amount": 999.0,  "ts": 1_716_200_003_000},
    {"customer": "alice", "amount": 340.0,  "ts": 1_716_200_005_000},
    {"customer": "dave",  "amount": 77.0,   "ts": 1_716_200_008_000},
    {"customer": "bob",   "amount": 210.0,  "ts": 1_716_200_012_000},
    {"customer": "eve",   "amount": 55.0,   "ts": 1_716_200_013_000},
    {"customer": "alice", "amount": 130.0,  "ts": 1_716_200_015_000},
    {"customer": "carol", "amount": 820.0,  "ts": 1_716_200_018_000},
    {"customer": "frank", "amount": 60.0,   "ts": 1_716_200_022_000},
]


def window_start(ts_ms: int) -> int:
    return (ts_ms // WINDOW_MS) * WINDOW_MS


def main() -> None:
    print("=== Enterprise 04 (Python): Kafka → 10s Tumbling Window ===")

    events = _collect_kafka(TIMEOUT_S)
    if not events:
        print("  no live Kafka data — using synthetic time-series events")
        events = SYNTHETIC_EVENTS

    print(f"  processing {len(events)} events…")

    # Assign events to windows.
    windows: dict[tuple, dict] = defaultdict(lambda: {"count": 0, "sum_amount": 0.0})
    watermark = 0
    late_count = 0

    for e in sorted(events, key=lambda x: x.get("ts", 0)):
        ts = e.get("ts", 0)
        customer = e.get("customer", "?")
        amount = float(e.get("amount", 0))

        wstart = window_start(ts)
        wend = wstart + WINDOW_MS
        key = (wstart, wend, customer)

        if ts < watermark - WATERMARK_LAG_MS:
            late_count += 1

        windows[key]["count"] += 1
        windows[key]["sum_amount"] += amount
        watermark = max(watermark, ts)

    # Format results.
    rows = []
    for (ws, we, customer), agg in sorted(windows.items()):
        rows.append({
            "window_start": ws,
            "window_end":   we,
            "customer":     customer,
            "order_count":  agg["count"],
            "total_amount": round(agg["sum_amount"], 2),
        })

    df = pd.DataFrame(rows).sort_values(["window_start", "total_amount"], ascending=[True, False])
    print("\n--- Tumbling window results (10s buckets) ---")
    print(df.to_string(index=False))

    summary = df.groupby("window_start").agg(
        customers=("customer", "nunique"),
        orders=("order_count", "sum"),
        revenue=("total_amount", "sum"),
    ).reset_index()
    print("\n--- Per-window summary ---")
    print(summary.to_string(index=False))

    if late_count:
        print(f"\n  late events (>{WATERMARK_LAG_MS}ms behind watermark): {late_count}")


def _collect_kafka(timeout_s: float) -> list[dict]:
    consumer = Consumer({
        "bootstrap.servers": BOOTSTRAP,
        "group.id": "krishiv-ent-py-04",
        "enable.auto.commit": True,
        "auto.offset.reset": "earliest",
    })
    consumer.subscribe(["orders"])
    events: list[dict] = []
    deadline = time.monotonic() + timeout_s
    try:
        while time.monotonic() < deadline:
            msg = consumer.poll(timeout=1.0)
            if msg is None:
                continue
            if msg.error():
                break
            try:
                events.append(json.loads(msg.value().decode("utf-8")))
            except Exception:
                continue
    finally:
        consumer.close()
    return events


if __name__ == "__main__":
    main()

File: examples/enterprise/rust/src/bin/ent_04_kafka_tumbling_window.rs
//! Enterprise 04 · Kafka → 10-second tumbling window → console — embedded mode
//!
//! Assigns order events (with an Int64 epoch-ms timestamp) to 10-second
//! tumbling windows, computing count + sum(amount) per (window, customer).
//!
//! Uses `execute_windowed_stream` from `krishiv` (re-exported from
//! `krishiv_runtime`) with `LocalWindowExecutionSpec`.
//!
//! Run:
//!   cargo run -p krishiv-enterprise-examples --bin ent_04_kafka_tumbling_window

use std::sync::Arc;

use anyhow::Result;
use arrow::array::{Float64Array, Int64Array, StringArray};
use arrow::datatypes::{DataType, Field, Schema};
use arrow::record_batch::RecordBatch;
use krishiv::{AggExpr, AggFunction};
use krishiv_runtime::{LocalWindowExecutionSpec, LocalWindowKind, execute_windowed_stream};

#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Enterprise 04: Tumbling Window (embedded) ===");

    // Events spanning three 10-second windows (base_ms, base+10s, base+20s).
    let base: i64 = 1_716_200_000_000;
    let schema = Arc::new(Schema::new(vec![
        Field::new("customer", DataType::Utf8,    false),
        Field::new("amount",   DataType::Float64, false),
        Field::new("ts",       DataType::Int64,   false),
    ]));

    let batch = RecordBatch::try_new(schema, vec![
        Arc::new(StringArray::from(vec![
            "alice", "bob",   "carol", "alice", "dave",
            "bob",   "eve",   "alice", "carol", "frank",
        ])),
        Arc::new(Float64Array::from(vec![
            120.0, 45.0, 999.0, 340.0, 77.0,
            210.0, 55.0, 130.0, 820.0, 60.0,
        ])),
        Arc::new(Int64Array::from(vec![
            base,           base + 1_000,
            base + 3_000,   base + 5_000,
            base + 8_000,   // window 0 (0..10s)
            base + 12_000,  base + 13_000,
            base + 15_000,  base + 18_000, // window 1 (10..20s)
            base + 22_000,                 // window 2 (20..30s)
        ])),
    ])?;

    let spec = LocalWindowExecutionSpec {
        key_column:      "customer".into(),
        key_column_type: "utf8".into(),
        event_time_column: "ts".into(),
        watermark_lag_ms: 2_000,
        window_kind:     LocalWindowKind::Tumbling,
        window_size_ms:  10_000,
        agg_exprs: vec![
            AggExpr {
                function: AggFunction::Count,
                input_column: String::new(),
                output_column: "order_count".into(),
            },
            AggExpr {
                function: AggFunction::Sum,
                input_column: "amount".into(),
                output_column: "total_amount".into(),
            },
        ],
        state_ttl_ms: None,
        source_watermark_lags: std::collections::HashMap::new(),
        source_id_column: None,
        window_timezone: None,
    };

    let output = execute_windowed_stream(vec![batch], &spec)?;

    println!("  {} result batches", output.len());
    let total_rows: usize = output.iter().map(|b: &RecordBatch| b.num_rows()).sum();

    if !output.is_empty() {
        let session = krishiv::Session::builder().build()?;
        session.register_record_batches("windows", output)?;
        let df = session.sql(
            "SELECT window_start_ms, window_end_ms, customer, order_count, total_amount \
             FROM windows ORDER BY window_start_ms, total_amount DESC"
        )?;
        println!("\n--- Tumbling window results (10s buckets) ---");
        println!("{}", df.collect()?.pretty()?);
    }

    println!("\n✓ {} window rows produced", total_rows);
    Ok(())
}
